[quant] QNNPACK Add deconvolution parameters

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37497 [quant] QNNPACK factoring out common structs from conv/deconv
* #36790 [quant] QNNPACK deconv
* #37405 [quant] QNNPACK deconvolution packing
* **#36716 [quant] QNNPACK Add deconvolution parameters**

Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

Commit:
[quant] Refactor QNNPACK conv for deconv

[ghstack-poisoned]

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"




[ghstack-poisoned]

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"




[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/src/deconv-run.cc:
```diff
@@ -0,0 +1,25 @@
+    const uint8_t output_zero_point,
+    const bool transpose,
+    uint8_t* output,
+    pthreadpool_t threadpool);
```
Can we move the changes in this file and the declaration of qnnpackDeConv over to the next PR?

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -190,5 +211,160 @@ struct conv_param_t {
+  /**
+   * @brief Constructor for initializing the deconvolution parameters.
+   */
+  deconv_param_t(
```
This looks exactly the same as conv_param_t to me. We should just reuse that.

z-a-f replied:
They are not exactly the same -- there are additional parameters in the deconv + the output shape is computed differently + the supported kernel type is only conv (no other kernels). Squashing the parameters together will require me to change the signature, which will require to change the files that already use the params. I added a TODO to merge the two, but I doin't think they should be part of this PR.

kimishpatel replied:
It seems to me that except for this parameter, `adjustment_dims`, everything else is the same. It should be ok to have this present for both conv/deconv. It could be renamed to `deconv_ adjustment_dims`. Can you elaborate a little bit on the change of signature comment and the subsequent change of the files comment you made?
I would like to minimize code duplication, if possible, especially as I implement per channel quant. 
It is ok if you want to do this as separate PR, but I would request you follow-up quickly on the merging TODO, otherwise we will just continue to add more duplicated code especially as I implement per channel quant.

z-a-f replied:
Will send a separate PR for the merging

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"




[ghstack-poisoned]

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"




[ghstack-poisoned]

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"




[ghstack-poisoned]

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/qnnpack_func.h:
```diff
@@ -107,6 +107,20 @@ enum pytorch_qnnp_status qnnpackConv(
     uint8_t* output,
     pthreadpool_t threadpool);
 
+enum pytorch_qnnp_status qnnpackDeConv(
```
Do we need this declaration in this PR?

supriyar commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -190,5 +211,160 @@ struct conv_param_t {
+  }
+};
+
+struct deconv_param_t {
```
The PR title says refactor conv but this adds a new struct for deconv. We aren't using this struct here, why don't we move it to the PR where we actually use it?

z-a-f replied:
The whole point of the refactor is to prep for the deconv struct, and add hte parameters. I will rename it

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -190,5 +211,160 @@ struct conv_param_t {
+      const std::array<uint32_t, 2> stride,
+      const std::array<uint32_t, 2> dil,
+      const std::array<uint32_t, 4> pd,
+      const std::array<uint32_t, 2> adj,
```
We can do better names. Perhaps trail names with _ to avoid confusion?

z-a-f replied:
These names are coming directly from the original QNNPACK conv_param_t naming. Besides, what can this be confused with? I don't mind changing the var names, but that calls for refactor of the qnnpack as a whole.

kimishpatel replied:
Just a more readability argument. Looking at pd does not tell me anything right away, or `dil`. I then have to look a few lines down. I was commenting from a practice perspective. I am okish if you wanna leave it like that but changing here doesn't have to be coupled with conv_params_t. That one can be done in a separate PR. I leave it upto you.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -8,9 +8,30 @@
+static inline size_t compute_output_dimension(
+    size_t input_dim,       // Input dimension
+    size_t pad_dim,         // Input padding
+    size_t adjustment_dim,  // Adjustment to the output dimension
```
What is adjustment_dim? Note that there is already conv util that can help you determine output shape here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ConvUtils.h. If possible let's use that?

z-a-f replied:
It doesn't have the deconv output shape.

kimishpatel replied:
Sounds good.

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -190,5 +211,160 @@ struct conv_param_t {
+  /**
+   * @brief Computes the output dimensions given a 2D input.
+   */
+  std::array<size_t, 2> compute_output_dims(
```
See the comment above for similar function in PT.

z-a-f replied:
The function in PT doesn't have the deconv shape

kimishpatel commented on aten/src/ATen/native/quantized/cpu/qnnpack/include/conv_utils.h:
```diff
@@ -190,5 +211,160 @@ struct conv_param_t {
+    const uint32_t input_padding_bottom = padding_dims[2];
+    const uint32_t input_padding_right = padding_dims[3];
+
+    group_input_channels = input_channels / groups;
```
We should add a check here for groups > 0 and input_channels % groups == 0? This is common to both conv and deconv. If you can merge them that would be good to, although I know we decided to defer this to a later PR. But this might just reduce the places that needs such fixing.

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]

Commit:
Update on "[quant] Refactor QNNPACK conv for deconv"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK Add deconvolution parameters"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK Add deconvolution parameters"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]

supriyar reviewed (approved):
LGTM

Commit:
Update on "[quant] QNNPACK Add deconvolution parameters"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]

Commit:
Update on "[quant] QNNPACK Add deconvolution parameters"


Differential Revision: [D21110112](https://our.internmc.facebook.com/intern/diff/D21110112)

[ghstack-poisoned]
